### PR TITLE
Try to undo `rbd lock/map` if `format/mount` fails in `Mounter.Setup/SetupAt` call.

### PR DESCRIPTION
**Pull request by cofyc**
**What this PR does / why we need it**:

Currently, RBD plugin does `rbd lock/map` and `format/mount` in one `Mounter.Setup` call. If `format/mount` fails, the `Mounter.Setup` fails, no `Unmouter.TearDown` will be executed, `rbd lock/map` cannot be undone.

The lock can be cleaned up when the image is to be locked next time because it's an orphaned lock (no watcher on the image). But `rbd map` will be impossible to be undone. The image will be mapped on the node forever, which will cause the PV cannot be deleted (still be used by a node). 

In this PR I fix it by undoing `rbd lock/map` immediately if `format/mount` fails. 

Of course, this is not a perfect solution, but it's the only way I can figure out. It reduces problems that may occur. It's better if we can split the `Mounter.Setup` process into `Attacher.Attach` and `Mounter.Setup`. But RBD plugin cannot implement `Attacher` interface currently, due to it cannot get secret string in `attacher.Attacher` call if user specified `secretName` under their namespace (do know which namespace to get the secret).

**Special notes for your reviewer**:

I had tried to implement `Attacher` interface, but found it's impossible (due the reason above). But I think this issue should be fixed. Pass Pod/User information in `Attacher.Attach` or add `secretNamespace` field in RBD volume spec? 

**Release note**:

```
NONE
```

/cc @rootfs

### CODE REVIEW

**Comment by rootfs** [comment]
@cofyc rbd advisory locks have limitations. Format failure is not the only issue. 

There are several proposals
- shifting to exclusive lock but disabling lock own transition. This works for rbd images with exclusive lock feature, but need latest kernel and rbd binary
- using rbd-nbd https://github.com/kubernetes/kubernetes/pull/38936. This requires rbd-nbd binary though.
- implementing rbd attach and detach support https://github.com/kubernetes/kubernetes/pull/33660. This requires rbd binary on controller image though.

If you can help any of the above, that'll be great.

File: pkg/volume/rbd/rbd_util.go [hunk 1]
```diff
@@ -317,6 +318,34 @@ func (util *RBDUtil) AttachDisk(b rbdMounter) error {
 	// mount it
 	if err = b.mounter.FormatAndMount(devicePath, globalPDPath, b.fsType, nil); err != nil {
 		err = fmt.Errorf("rbd: failed to mount rbd volume %s [%s] to %s, error %v", devicePath, b.fsType, globalPDPath, err)
+		// Currently, RBD plugin does `rbd lock/map` and `format/mount` in one
```
Comment (rootfs) [inline pkg/volume/rbd/rbd_util.go]: this shares lots with `DetachDisk`, I would just reuse `DetachDisk` here

Comment (cofyc) [inline pkg/volume/rbd/rbd_util.go]: @rootfs 
`DetachDisk` gets `devicePath` from mount point and need to report error if it fails to unmount the point. I cannot call it directly. I added a new subroutine `RBDUtil.detachDevice`.

**Comment by cofyc** [comment]
hi, @rootfs I will have a look at issues you mentioned.

**Comment by cofyc** [comment]
@rootfs 

Hi, I submit a [PR](https://github.com/kubernetes/kubernetes/pull/51608) to continue your work https://github.com/kubernetes/kubernetes/pull/33660. Implementing Attacher/Detacher is best way.

Don't know why you discontinue https://github.com/kubernetes/kubernetes/pull/33660 ...

Non-RBD changes are almost same.

**Pull request closed without merging**